community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/IndexTxStateUpdater.java

```diff
@@ -77,18 +79,15 @@ public void onLabelChange( KernelStatement state, int labelId, NodeItem node, La
             int[] indexPropertyIds = index.schema().getPropertyIds();
             if ( nodeHasIndexProperties( nodePropertyIds, indexPropertyIds ) )
             {
-                ValueTuple values = getValueTuple( state, node, indexPropertyIds );
+                Value[] values = getValueTuple( state, node, indexPropertyIds );
                 if ( changeType == LabelChangeType.ADDED_LABEL )
                 {
-                    for ( int i = 0; i < values.size(); i++ )
-                    {
-                        Validators.INDEX_VALUE_VALIDATOR.validate( values.valueAt(i) );
-                    }
-                    state.txState().indexDoUpdateEntry( index.schema(), node.id(), null, values );
+                    indexingService.validateBeforeCommit( index.schema(), values );
```

Comment:
I don't really like that we go all the way down to just schema here, this assumes that we only have one active index provider. It would be better if we were to fetch some data structure that also has knowledge of which provider it belongs to, and then use that information to get the provider to validate with. If that is too much work, at least pass down the entire index descriptor, so that the validateBeforeCommit can use getProviderFor to get the provider in 3.5.

Reply:
This whole PR is for allowing individual indexes (basically individual providers) to have their own validation, which it accomplishes and verifies through tests. If you could point me to where this single provider assumption is made I'd be happy to make changes to allow for it, but as far as I can see this works for multiple active providers.

Reply:
Upon closer inspection, I guess this makes sense for now. I'd rather see us use indexid or something here, but that isn't readily available.